### PR TITLE
hack(gatsby-remark-prismjs): split multiline plain-text tokens into multiple spans

### LINKED ISSUE #8058: code line highlighting buggy
https://twitter.com/rchrdnsh/status/1039529095372009472

### PR DESCRIPTION
**Pull request by pieh**
First of all - I'm ashamed of my hack (regex + html = 💔 )!

Now let me try to explain the problem:
After https://github.com/gatsbyjs/gatsby/pull/7342 jsx code blocks get some extra tokens to not highlight some parts of code (plain-text tokens).

```
<ul>
  <li>text</li>
</ul>
```

will get something like:

```
[regular tokens for `<ul>`]<span class="token plain-text"> 
  </span>[regular tokens for `<li>` part]<span class="token plain-text">text</span>[regular tokens for `</li>` part]<span class="token plain-text">
</span>[regular tokens for `</ul>` part]
```

The way we highlight code lines is we split html string by `\n` and wrap lines we want to highlight with highlight span - so if we would want to highlight 2nd line (`  <li>text</li>`) we would get:

```
[regular tokens for `<ul>`]<span class="token plain-text"> 
<span class="gatsby-highlight-code-line">  </span>[regular tokens for `<li>` part]<span class="token plain-text">text</span>[regular tokens for `</li>` part]<span class="token plain-text"></span>
</span>[regular tokens for `</ul>` part]
```

so what happens here is highlight span gets closed by closing tag that previously belonged to plain-text span, and because highlight is `display: block`, we render empty highlighted line and content gets pushed to newline (without any intendation) - see screenshots in https://github.com/gatsbyjs/gatsby/issues/8058#issuecomment-420449176

My hacky workaround is to not allow plain-text spans to contain `\n` and instead create separate plain-text spans and move `\n` outside of it - this is very hacky, but it works - ideally this could be fixed in `prismjs`, but just to fix our plugin I think we need this (or something like this). 

I probably will need to add some tests here, just not sure what's the best way to do that yet as snapshots won't really cut it here I think.

fixes #8058

### CODE REVIEW

**Comment by fk** [comment]
[<3](https://www.youtube.com/watch?v=WoZP18U5VyY) [Shake](https://www.youtube.com/watch?v=i7OEvo-GjUg) [the](https://www.youtube.com/watch?v=OwC5ePDNkp0) [shame](https://www.youtube.com/watch?v=DJhGNcwwuME)! 😋 

Super glad you took care of this — just seeing [those first ten lines](https://github.com/PrismJS/prism/blob/master/components/prism-jsx.js) made me __kthnxbai__! 🤗 🙏 

Re:fixing this upstream, I wondered if _not_ closing all token `span`s of a LOC is an expected thing in prism/for plain text nodes … but I clearly don't have the brains to even get close to understand all that stuff. Regex nightmares! 😰😅

File: packages/gatsby-remark-prismjs/src/highlight-code.js [hunk 1]
```diff
@@ -3,6 +3,8 @@ const _ = require(`lodash`)
 
 const loadPrismLanguage = require(`./load-prism-language`)
 
+const plainTextWithLFTest = /<span class="token plain-text">[^<]*\n[^<]*<\/span>/g
```
Comment (DSchau) [inline packages/gatsby-remark-prismjs/src/highlight-code.js]: I'll taking hacky and working over the alternative every day of the week!

**Comment by pieh** [comment]
> Re:fixing this upstream, I wondered if _not_ closing all token `span`s of a LOC is an expected thing in prism/for plain text nodes … but I clearly don't have the brains to even get close to understand all that stuff. Regex nightmares! 😰😅

We don't need to fix issue ourselves upstream - first step would be to create issue and ask if this is something that they consider a problem, because it seems like this is problem only in scenarios like ours when we manipulate html string ourselves. They implement highlighting differently - i think it's done client side by creating absolute positioned highlights so they don't change html.

**Comment by DSchau** [comment]
@pieh do you mind if I merge and release this? I know it's "hacky" but if it works and saves people some headaches, I think it's worth it!

**Comment by DSchau** [comment]
```
Successfully published:
 - gatsby-remark-prismjs@3.0.2
```

Thanks @pieh!

**Comment by fk** [comment]
LOL yeah I remember that stackoverflow answer 👍 "all is lost the pony the pony" 😸

**Pull request merged**